File: server/datasets/python_source-code/Multiple-A1/unique1_binary_search.py
"""
Binary Search Suite - Submission by Isabelle
Approach: Defensive, type-annotated binary search toolkit returning SearchResult objects,
including first occurrence, last occurrence, and range counting with validation.
"""

from dataclasses import dataclass
from typing import List, Optional, Tuple


@dataclass
class SearchResult:
    index: Optional[int]
    comparisons: int
    found: bool


def validate_sorted(arr: List[int]) -> bool:
    return all(arr[i] <= arr[i + 1] for i in range(len(arr) - 1))


def binary_search(arr: List[int], target: int) -> SearchResult:
    if not validate_sorted(arr):
        raise ValueError("Array must be sorted")
    lo, hi = 0, len(arr) - 1
    comps = 0
    while lo <= hi:
        mid = (lo + hi) // 2
        comps += 1
        if arr[mid] == target:
            return SearchResult(index=mid, comparisons=comps, found=True)
        elif arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return SearchResult(index=None, comparisons=comps, found=False)


def find_first_occurrence(arr: List[int], target: int) -> Optional[int]:
    lo, hi = 0, len(arr) - 1
    first_idx = None
    while lo <= hi:
        mid = (lo + hi) // 2
        if arr[mid] == target:
            first_idx = mid
            hi = mid - 1
        elif arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return first_idx


def find_last_occurrence(arr: List[int], target: int) -> Optional[int]:
    lo, hi = 0, len(arr) - 1
    last_idx = None
    while lo <= hi:
        mid = (lo + hi) // 2
        if arr[mid] == target:
            last_idx = mid
            lo = mid + 1
        elif arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return last_idx


def count_frequency(arr: List[int], target: int) -> int:
    first = find_first_occurrence(arr, target)
    if first is None:
        return 0
    last = find_last_occurrence(arr, target)
    return last - first + 1


def main():
    sample = [2, 5, 8, 12, 12, 12, 23, 38, 45, 56, 72, 91]
    res = binary_search(sample, 12)
    print("Found:", res.found, "at index:", res.index)
    print("Frequency of 12:", count_frequency(sample, 12))


if __name__ == "__main__":
    main()

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_a1/unique1_binary_search.java
/*
 * Binary Search - Submission by Isabelle
 * Approach: Defensive, fully type-hinted implementation that returns a
 * small SearchResult object (index + comparison count) instead of a
 * bare int, and validates that the input is actually sorted first.
 * This is the "unique" baseline for this scenario.
 */

class SearchResult {
    Integer index;
    int comparisons;

    public SearchResult(Integer index, int comparisons) {
        this.index = index;
        this.comparisons = comparisons;
    }

    public boolean isFound() {
        return this.index != null;
    }
}

class BinarySearch {
    public static boolean isSorted(int[] arr) {
        for (int i = 0; i < arr.length - 1; i++) {
            if (arr[i] > arr[i + 1]) {
                return false;
            }
        }
        return true;
    }

    public static SearchResult binarySearch(int[] arr, int target) {
        if (!isSorted(arr)) {
            throw new IllegalArgumentException("binary_search requires a sorted list");
        }

        int lo = 0;
        int hi = arr.length - 1;
        int comparisons = 0;

        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            comparisons++;

            if (arr[mid] == target) {
                return new SearchResult(mid, comparisons);
            } else if (arr[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return new SearchResult(null, comparisons);
    }

    public static void main(String[] args) {
        int[] arr = {2, 5, 8, 12, 16, 23, 38, 45, 56, 72, 91};
        int target = 23;

        SearchResult result = binarySearch(arr, target);
        if (result.isFound()) {
            System.out.println("Found " + target + " at index " + result.index + " (" + result.comparisons + " comparisons)");
        } else {
            System.out.println(target + " not found (" + result.comparisons + " comparisons)");
        }
    }
}
